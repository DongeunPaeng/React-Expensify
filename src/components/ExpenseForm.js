import React from "react";
import { DatePicker } from "rsuite";
import { DateTime } from "luxon";

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense.description ? props.expense.description : "",
      note: props.expense.note ? props.expense.note : "",
      amount: props.expense.amount ? props.expense.amount : "",
      createdAt: props.expense.createdAt
        ? DateTime.fromMillis(props.expense.createdAt).ts
        : DateTime.local(),
      message: undefined
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    this.setState(() => ({ amount }));
  };

  onDateChange = date => {
    const createdAt = DateTime.fromISO(date.toISOString()).ts;
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        message: "please provide description and amount"
      }));
    } else {
      this.setState(() => ({ message: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseInt(this.state.amount),
        createdAt: this.state.createdAt,
        notes: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.message && <p>{this.state.message}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="number"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <DatePicker onChange={this.onDateChange} />
          <textarea
            placeholder="type something(optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
