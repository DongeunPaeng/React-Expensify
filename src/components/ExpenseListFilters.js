import React from "react";
import { connect } from "react-redux";
import {
  setStartDate,
  setEndDate,
  setTextFilter,
  sortByDate,
  sortByAmount
} from "../actions/filters";
import { DateRangePicker } from "rsuite";
import { DateTime } from "luxon";

export class ExpenseListFilters extends React.Component {
  onTextChange = e => {
    this.props.setTextFilter(e.target.value); // the dispatch function receives actions, which is defined in actions folder by actionCreators(very important concept).
  };
  onSortChange = e => {
    if (e.target.value === "date") {
      this.props.sortByDate(); // this dispatch will send the given action to reducer and ask the reducer to return updated state.
    } else if (e.target.value === "amount") {
      this.props.sortByAmount();
    }
  };
  onDateChange = date => {
    this.props.setStartDate(DateTime.fromISO(date[0].toISOString()).ts);
    this.props.setEndDate(DateTime.fromISO(date[1].toISOString()).ts);
  };

  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              className="text-input"
              placeholder="Search Expenses"
              type="text"
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker onChange={this.onDateChange} />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters
});

const mapDispatchToProps = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);

// connect does this thing: it connects the React component with Redux store, by receiving mapStateToProps and mapDispatchToProps functions. Thus, the React component receives 'store' as the props.
// mapStateToProps does this thing: it enables the React component to reach Redux store and bring any data from it with the form of 'props.'
