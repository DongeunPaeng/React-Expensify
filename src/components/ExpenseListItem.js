import React from "react";
import { Link } from "react-router-dom";
import { DateTime } from "luxon";
import numeral from "numeral";

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <Link className="list-item" to={`/edit/${id}`}>
    <div>
      <h3 className="list-item__title">{description}</h3>
      <span className="list-item__sub-title">
        {DateTime.fromMillis(createdAt).toLocaleString()}
      </span>
    </div>
    <h3 className="list-item__data">{numeral(amount).format("0,0")}Ïõê</h3>
  </Link>
);

export default ExpenseListItem;
