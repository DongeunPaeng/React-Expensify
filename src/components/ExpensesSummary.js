import React from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import selectExpenses from "../selectors/expenses";
import selectExpensesTotal from "../selectors/expenses-total";
import numeral from "numeral";

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => (
  <div className="page-header">
    <div className="content-container">
      <h1 className="page-header__title">
        There are <span>{numeral(expenseCount).format("0,0")}</span> expenses totalling{" "}
        KRW <span>{numeral(expensesTotal).format("0,0")}</span>
      </h1>
      <div className="page-header__actions">
        <Link to="/create" className="button">
          Add Expense
        </Link>
      </div>
    </div>
  </div>
);

const mapStateToProps = state => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
