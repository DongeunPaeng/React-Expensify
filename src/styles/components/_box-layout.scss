.box-layout {
  align-items: center;
  background: url("../../public/images/bg.jpg");
  background-size: cover;
  justify-content: center;
  display: flex;
  height: 100vh;
  width: 100vw;
}

.box-layout__box {
  background: fade-out(white, 0.15);
  text-align: center;
  padding: $l-size $m-size;
  border-radius: 3px;
  width: 30rem;
}

.box-layout__title {
  margin: 0 0 $m-size 0;
  line-height: 1;
}
