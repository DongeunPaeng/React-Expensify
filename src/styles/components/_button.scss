.button {
  display: inline-block;
  border: none;
  text-decoration: none;
  background: $blue;
  color: white;
  line-height: 1;
  font-size: $font-size-large;
  font-weight: 300;
  padding: $s-size $m-size;
  margin-top: 1rem;
}

.button--link {
  background: none;
}

.button--secondary {
  background: #888;
}
