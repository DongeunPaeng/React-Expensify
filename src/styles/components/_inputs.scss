.text-input {
  border: 1px solid #e5e5ea;
  border-radius: 6px;
  height: 36px;
  color: #8e8e93;
  font-size: 14px;
  font-weight: 300;
  padding: $s-size;
}

.select {
  @extend .text-input;
}

.textarea {
  @extend .text-input;
  height: 10rem;
}
