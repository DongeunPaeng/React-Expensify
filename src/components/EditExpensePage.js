import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {
  onSubmit = expense => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push("/"); // 이게 뭐더라?
  };
  onRemove = () => {
    this.props.startRemoveExpense(this.props.expense.id);
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
          <button className="button button--secondary" onClick={this.onRemove}>Remove Expense</button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = dispatch => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: id => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);

// If there is nothing to read from the Redux store, I don't need to put mapStateToProps argument here in the connect function.
// To understand this better, I need to know what role mapStateToProps has.
// The main role of mapStateToProps is "to subscribe" the Redux state. It is unknown if it is needed to pass Redux state to React component, too.
// Like mapStateToProps passes state as props to React component, mapDispatchToProps passes dispatch as props. So, unless there is something to specify like state.expenses, dispatch.removeItem, I can just make it null.
// The important concept here is if I need to specify state.xxx or dispatch.xxx, which determines whether I need to pass mapState/DispatchToProps arguments.
