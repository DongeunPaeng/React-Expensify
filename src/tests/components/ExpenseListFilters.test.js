import React from "react";
import { shallow } from "enzyme";
import { ExpenseListFilters } from "../../components/ExpenseListFilters";
import { filters, altFilters } from "../fixtures/filters";
import { DateTime } from "luxon";

let setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate, wrapper;

beforeEach(() => {
  setTextFilter = jest.fn();
  sortByDate = jest.fn();
  sortByAmount = jest.fn();
  setStartDate = jest.fn();
  setEndDate = jest.fn();
  wrapper = shallow(
    <ExpenseListFilters
      filters={filters}
      setTextFilter={setTextFilter}
      sortByDate={sortByDate}
      sortByAmount={sortByAmount}
      setStartDate={setStartDate}
      setEndDate={setEndDate}
    />
  );
});

test("should render ExpenseListFilters correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseListFilters alt data correctly", () => {
  wrapper.setProps({
    filters: altFilters
  });
  expect(wrapper).toMatchSnapshot();
});

test("should handle text change", () => {
  const value = "rent";
  wrapper.find("input").simulate("change", {
    target: { value }
  });
  expect(setTextFilter).toHaveBeenLastCalledWith(value);
});

test("should sort by date", () => {
  const value = "date";
  wrapper.setProps({
    filters: altFilters
  });
  wrapper.find("select").simulate("change", {
    target: { value }
  });
  expect(sortByDate).toHaveBeenCalled();
});

test("should sort by amount", () => {
  const value = "amount";
  wrapper.find("select").simulate("change", {
    target: { value }
  });
  expect(sortByAmount).toHaveBeenCalled();
});

test("should handle date change", () => {
  const startDate = new Date(2020, 10, 10);
  const luxonStartDate = DateTime.fromISO(startDate.toISOString()).ts;
  const endDate = new Date(2020, 10, 11);
  const luxonEndDate = DateTime.fromISO(endDate.toISOString()).ts;
  const date = [startDate, endDate];
  wrapper
    .find("withLocale(defaultProps(DateRangePicker))")
    .simulate("change", date);
  expect(setStartDate).toHaveBeenLastCalledWith(luxonStartDate);
  expect(setEndDate).toHaveBeenLastCalledWith(luxonEndDate);
});
