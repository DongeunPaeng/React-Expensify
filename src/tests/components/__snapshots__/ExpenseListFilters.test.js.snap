// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListFilters alt data correctly 1`] = `
<div
  className="content-container"
>
  <div
    className="input-group"
  >
    <div
      className="input-group__item"
    >
      <input
        className="text-input"
        onChange={[Function]}
        placeholder="Search Expenses"
        type="text"
        value="bills"
      />
    </div>
    <div
      className="input-group__item"
    >
      <select
        className="select"
        onChange={[Function]}
        value="amount"
      >
        <option
          value="date"
        >
          Date
        </option>
        <option
          value="amount"
        >
          Amount
        </option>
      </select>
    </div>
    <div
      className="input-group__item"
    >
      <withLocale(defaultProps(DateRangePicker))
        onChange={[Function]}
      />
    </div>
  </div>
</div>
`;

exports[`should render ExpenseListFilters correctly 1`] = `
<div
  className="content-container"
>
  <div
    className="input-group"
  >
    <div
      className="input-group__item"
    >
      <input
        className="text-input"
        onChange={[Function]}
        placeholder="Search Expenses"
        type="text"
        value=""
      />
    </div>
    <div
      className="input-group__item"
    >
      <select
        className="select"
        onChange={[Function]}
        value="date"
      >
        <option
          value="date"
        >
          Date
        </option>
        <option
          value="amount"
        >
          Amount
        </option>
      </select>
    </div>
    <div
      className="input-group__item"
    >
      <withLocale(defaultProps(DateRangePicker))
        onChange={[Function]}
      />
    </div>
  </div>
</div>
`;
